js/ai/src/resource.ts

```diff
@@ -121,7 +126,25 @@ export function defineResource(
           message: `input ${input} did not match template ${uri}`,
         });
       }
-      return await fn(templateMatch, ctx);
+      const parts = await fn(input, ctx);
+      parts.map((p) => {
+        if (!p.metadata) {
+          p.metadata = {};
+        }
+        if (p.metadata?.resource) {
+          (p.metadata as any).resource.parent = {
+            uri: input,
+            ...(p.metadata as any).resource.parent,
+          };
+        } else {
+          (p.metadata as any).resource = {
+            name: opts.name,
+            uri: input,
```

Comment:
I think the important metadata is `uri` and `template` (I'm ambivalent about including name, but do think template should be included if it exists).